utils/sysfs/sysfs.go

```diff
@@ -335,25 +341,143 @@ func (fs *realSysFs) GetSystemUUID() (string, error) {
+	onlineVal, err := ioutil.ReadFile(onlinePath)
+	// First, make sure that no error occurred.
 	if err != nil {
-		pathErr, ok := err.(*os.PathError)
-		if ok {
-			if errors.Is(pathErr.Unwrap(), os.ErrNotExist) && isZeroCPU(dir) {
+		// If error occurred and is os.PathError then it might be os.ErrNotExists.
+		pathErr, isPathError := err.(*os.PathError)
+		if isPathError {
+			if !os.IsNotExist(pathErr.Unwrap()) {
+				// If online file exists but reading failed we assume that CPU is offline.
+				klog.V(2).Infof("Cannot determine CPU %s online state, assuming offline", cpuPath)
+				return false
+			}
+			// If online file does not exists then we should check one more location.
+			klog.V(2).Infof("%s not found, checking %s directory", onlinePath, filepath.Join(cpuPath, "hotplug"))
+			if isOnlineHotplug(cpuPath) {
 				return true
 			}
+			klog.V(2).Infof("unable to determine hotplug state of %s CPU; assuming that CPU hotplug is disabled", cpuPath)
+			return false
 		}
-		klog.Warningf("unable to read %s: %s", cpuPath, err.Error())
+	}
+	// Check if online file value is 1.
+	onlineVal = bytes.TrimSpace(onlineVal)
+	if len(onlineVal) == 0 || onlineVal[0] != isOnline {
+		klog.V(2).Infof("CPU %s is offline", cpuPath)
+		return false
+	}
+	return true
+}
+
+func isx86() bool {
+	uts := unix.Utsname{}
+	err := uname(&uts)
+	if err != nil {
 		return false
 	}
-	trimmed := bytes.TrimSpace(content)
-	return len(trimmed) == 1 && trimmed[0] == 49
+	return uts.Machine[0] == 120 && uts.Machine[1] == 56 && uts.Machine[2] == 54
 }
 
 func isZeroCPU(dir string) bool {
 	regex := regexp.MustCompile("cpu([0-9]*)")
 	matches := regex.FindStringSubmatch(dir)
 	return len(matches) == 2 && matches[1] == "0"
 }
+
+// isOnlineHotplug checks if CPU state matches online state from /sys/devices/system/cpu/hotplug/states.
+// It allows to determine if CPU is online on some ARM platforms such as RPi4.
+// See: https://www.kernel.org/doc/html/latest/core-api/cpu_hotplug.html#testing-of-hotplug-states
+func isOnlineHotplug(cpuPath string) bool {
+	desiredState, err := getOnlineStateID(cpuPath)
+	if err != nil {
+		klog.V(4).Infof("Cannot determine online state ID for CPU %s: %s", cpuPath, err)
+		return false
+	}
+	hotplugState := filepath.Join(cpuPath, "hotplug", "state")
+	state, err := ioutil.ReadFile(hotplugState)
+	if err != nil {
+		klog.V(4).Infof("Cannot read %s file: %s", hotplugState, err)
+		return false
+	}
+	state = bytes.TrimSpace(state)
+	realState, err := strconv.Atoi(string(state))
+	if err != nil {
+		klog.V(4).Infof("Cannot convert %s to int: %s", state, err)
+		return false
+	}
+	return realState == desiredState
+}
+
+// getOnlineStateID looks for line similar to `206: online` from /sys/devices/system/cpu/hotplug/states
+// and extracts number from the beginning.
+// See: https://www.kernel.org/doc/html/latest/core-api/cpu_hotplug.html#testing-of-hotplug-states
+func getOnlineStateID(cpuPath string) (int, error) {
```

Comment:
I think that it would make sense to cache result of this function: it can't change when system is running, I believe. Do you think that using package-level variable to cache the output makes sense?

Reply:
if it can't change, I guess doesn't hurt to cache it. re using package level variable, I think it's fine, e.g. see this existing example: https://github.com/google/cadvisor/blob/master/machine/machine.go#L53